Update Major upgrades cloud article (#8069)
umbraco/UmbracoDocs: umbraco-cloud/optimize-and-maintain-your-site/manage-product-upgrades/product-upgrades/major-upgrades.md

# Major Upgrades

> A major Umbraco upgrade is more involved than a minor or patch update. Customers who attempt it without understanding the full process often encounter unexpected site downtime. Read this article before touching any environment.

Reviewer: This sounds very "dire", like "things WILL absolutely go wrong" 😅, and even though it might be true, it's not a good look for the product, and it makes us look pretty bad.

How about this instead:

```suggestion
A major Umbraco upgrade is more involved than a minor or patch update. Read this article before starting the upgrade process to avoid any unexpected site downtime.
```

## Why major upgrades are different

> A major upgrade involves a database schema migration, the structure of the database changes between major versions. On Umbraco Cloud, this migration may not run automatically during a standard cloud deployment. When this occurs, you will need to run the database migration locally, where your machine has full control over the process. Once migrated locally via the connection string, verify the backoffice loads successfully in the cloud.

Reviewer: I don't understand the last sentence, so I've removed it. It sounded more like a step, and will hopefully be repeated as a step further down the guide 🤞 

```suggestion
A major upgrade involves a database schema migration, as the structure of the database changes between major versions. On Umbraco Cloud, this migration may not run automatically during a standard cloud deployment. When this occurs, you will need to run the database migration locally, where your machine has full control over the process.
```

Author: Yup, it's more of a setup step and there below so okay to remove :)

> A major upgrade involves a database schema migration, the structure of the database changes between major versions. On Umbraco Cloud, this migration may not run automatically during a standard cloud deployment. When this occurs, you will need to run the database migration locally, where your machine has full control over the process. Once migrated locally via the connection string, verify the backoffice loads successfully in the cloud.

Reviewer 2: If the database is the bottleneck then why do they need to run it from local? If it is due to ongoing traffic then swapping the DNS to staging should be enough?

Not sure I understand the reasoning for this complicated flow

Author: I totally get why this looks overly complicated at first glance, but the bottleneck isn't live user traffic; it's infrastructure resource allocation. Swapping DNS to Staging doesn't solve this, as the Staging environment is subject to the same hosting limitations.

Would it clarify if I update the text to:

## Why major upgrades are different

A major upgrade involves an intensive database schema migration where the structural layout of your database changes.

Unlike normal day-to-day operations, running a major structural data migration requires a significant temporary spike in raw CPU processing power and execution time. Because Umbraco Cloud web hosting containers are highly optimized to serve cached web traffic rather than execute long-running infrastructure modifications, a direct cloud deployment will frequently hit a hardware resource ceiling. This causes the environment to time out, resulting in boot failures or 503 errors.

To bypass this infrastructure constraint, your local machine must act as the high-powered execution engine. By configuring your local project files with a secure Cloud Connection String, your local computer provides the necessary CPU and RAM overhead to securely process the schema migration directly against the cloud database. Once this remote orchestration finishes, the data stays securely in the cloud, and you simply refresh your browser to verify that the backoffice loads successfully.

Reviewer 2: Alright I understand what you are trying to say here and why. 
I would reword it to sound less certain that things will fail though. It is not a certainty that the site will not be able to run these updates, it all depend on the types of migrations and the plan and traffic your site has.

![Umbraco Cloud Major Version Upgrade Roadmap](../../../.gitbook/assets/umbraco_upgrade_flow_process.svg)

{% hint style="warning" %}
**Do not perform major version upgrades directly in Umbraco Cloud.**

Major upgrades involve significant database schema changes. Running these migrations directly on Cloud can cause boot failures and 503 errors. These incomplete migrations are often difficult to diagnose.

Always perform the database migration locally first, verify that the backoffice loads successfully, and then deploy the upgraded project to Cloud. The steps below follow this recommended approach.
{% endhint %}

1. Upgrade the Dev environment by cloning it locally, running the upgrade, and push the upgraded code back up.
2. Deploy to Staging. The code will deploy but the database migration may not complete automatically. This is expected behaviour, not a bug.
3. Perform the Staging database migration locally by cloning Staging, connecting to its cloud database via a connection string, and running the project locally. Your local machine performs the migration against the cloud database.

> 4. Validate Staging thoroughly. Move your custom hostnames from Live to Staging so you can test on real traffic while Live remains untouched.

Reviewer 2: This makes the assumption that staging is a copy of live, that is mostly not the case even if it started out that way. It is also common to not have things like order flows pointing to the real payment provider from staging.

I understand the intent here, but in most cases I would argue it is better to kill traffic to the live site and put up a maintenance page somewhere than to point it to a staging site and assume it just works.

Author: See reply below.

5. Repeat the same local database migration process for Live.

> 6. Move hostnames back to Live. All environments are now upgraded.

Reviewer: I thought these were the actual steps, but they are actually just highlighting each step, or? 🤔 

If the user isn't supposed to follow these steps, but continue down to the actual steps, this list needs to change.

I'd suggest using an unordered list, and removing all the "details" - and ideally adding a disclaimer that these aren't the actual steps 😅 

```suggestion
The following list outlines the upgrade process. Find the detailed upgrade steps further down the article.

* Upgrade the left-most (Dev) environment.
* Deploy to the next environment (Staging).
* Perform the Staging database migration locally.
* Validate Staging thoroughly.
* Repeat the same local database migration process for Live.
* Move hostnames back to Live.
```

> ## Prerequisites

Reviewer: Is the Staging environment required for this guide? If so, that should be mentioned somewhere - ideally under Prerequisites.

Author: Yep. Staging is required because most of the customers/users have at least 3 environments. It also helps describe the process better for customers with many environments. Will add it under prereqs.

If the issue persists and your project has CDN Caching and Optimization enabled, cached responses may interfere with license validation. In this case, try purging the cache and reloading the site.
{% endhint %}

> 5. Verify the project runs locally without errors.

Reviewer: ```suggestion
4. Verify the project runs locally without errors.
```

You can choose between two approaches based on your needs:

* **With content freeze** (recommended for larger sites) - involves a more detailed upgrade process but helps reduce downtime on your live website.
* **Without content freeze** - provides a more straightforward process that may result in longer downtime on your live website.

{% tabs %}
{% tab title="With content freeze (recommended for larger sites)" %}
{% hint style="info" %}
The following steps involve setting a **content-freeze** period on the project. It is recommended to coordinate this with your content editors before moving forward.
{% endhint %}

> 1. Delete your existing Staging environment.

Reviewer: Should we keep the naming "general", or is it OK to use Staging and Dev?
If no, I'd suggest updating steps 1 and 2 here back to what they were before.

```suggestion
1. Delete any environments between your left-most (Dev) and production environments.
```

Author: I think we should use generic environment names in this article to avoid ambiguity. Without the general names, the **Step 5: Deploy the upgrade to Production** section will get too difficult to understand as to which environment we are referring to delete, recreate and clone.

> 2. Recreate the Staging environment as a direct clone from your production (Live) environment. This guarantees Staging begins as a clean environment with identical data.

Reviewer: It will always be a direct clone of the product environment:

```suggestion
2. Recreate the Staging environment to ensure a clean environment in sync with production/Live.
```

> 3. Initiate **content-freeze**. Ask editors to stop all backoffice work.

Reviewer 2: This needs more explanation, what if they use Umbraco Commerce? Then even if they have a content freeze they could loose out on orders being made.
The "content freeze" only works if they do not rely on persisting anything based on user interaction in the fronted, any member content, orders, analytics, scheduled jobs, etc stored in the db would be lost.

Author: That is a really good point, @jemayn I could add a Note stating, "If your site handles transactional frontend data (such as **Umbraco Commerce** orders, public member registrations, or user submissions), a standard backoffice content freeze will not stop users from interacting with the front end. Overwriting or restoring databases down the pipeline during active user traffic will result in **permanent data loss**."

Do we have a workaround for this @meyntony @jemayn? Will an extra line in the Note like below help or should they reach out to support?
For transactional sites, do not sync or overwrite databases down the pipeline. Instead, deploy the codebase directly and use a full **Application Maintenance Page** to pause frontend traffic while you run the Live local database migration loop.

Reviewer 2: It is really hard to give users a good flow that works for everyone. In my experience setting a maintenance window and diverting traffic is the safest option, but some clients will refuse that option. 
I think what you write here makes sense, we just need to make sure that we highlight the downsides of doing this and the data inconsistencies that may happen even with a content freeze - for some sites those inconsistencies will not matter, for others they are a dealbreaker.

> 4. Restore the production [database](../../../build-and-customize-your-solution/set-up-your-project/databases/backups.md) and [media](../../../build-and-customize-your-solution/handle-deployments-and-environments/deployment/restoring-content/) to Staging so the content is identical.

Reviewer 2: Restoring the db can be a very slow process if it contains a lot of data. I think we should flag that

Author: Will add a Note here.

5. Deploy the upgrade from the left-most (Dev) environment to Staging.

{% hint style="warning" %}

> The deployment to Staging may not automatically complete the database migration. If the site does not boot correctly after deployment, this is expected. Proceed to the local database migration step below rather than retrying the deployment.

Reviewer 2: Why is this the case?
